google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1987,58 +1987,67 @@ Blob create(
   Bucket update(BucketInfo bucketInfo, BucketTargetOption... options);
 
   /**
-   * Updates blob information. Original metadata are merged with metadata in the provided {@code
-   * blobInfo}. To replace metadata instead you first have to unset them. Unsetting metadata can be
-   * done by setting the provided {@code blobInfo}'s metadata to {@code null}. Accepts an optional
-   * userProject {@link BlobTargetOption} option which defines the project id to assign operational
-   * costs.
+   * Updates the blob properties if the preconditions specified by {@code options} are met. The
+   * property update works as described in {@link #update(BlobInfo)}.
    *
-   * <p>Example of udating a blob, only if the blob's metageneration matches a value, otherwise a
-   * {@link StorageException} is thrown.
+   * <p>{@code options} parameter allows to specify the preconditions for applying the update. E.g.
+   * update of the blob properties might be required only if the properties have not been updated
+   * externally. {@code StorageException} with the code {@code 412} is thrown if preconditions fail.
+   *
+   * <p>Example of updating the content type only if the properties are not updated externally:
    *
    * <pre>{@code
-   * String bucketName = "my-unique-bucket";
-   * String blobName = "my-blob-name";
-   * Blob blob = storage.get(bucketName, blobName);
-   * BlobInfo updatedInfo = blob.toBuilder().setContentType("text/plain").build();
-   * storage.update(updatedInfo, BlobTargetOption.metagenerationMatch());
+   * BlobId blobId = BlobId.of(bucketName, blobName);
+   * BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("text/plain").build();
+   * Blob blob = storage.create(blobInfo);
+   *
+   * doSomething();
+   *
+   * BlobInfo update = blob.toBuilder().setContentType("multipart/form-data").build();
+   * Storage.BlobTargetOption option = Storage.BlobTargetOption.metagenerationMatch();
+   * try {
+   *   storage.update(update, option);
+   * } catch (StorageException e) {
+   *   if (e.getCode() == 412) {
+   *     // the properties were updated externally
+   *   } else {
+   *     throw e;
+   *   }
+   * }
    * }</pre>
    *
+   * @param blobInfo information to update
+   * @param options preconditions to apply the update
    * @return the updated blob
-   * @throws StorageException upon failure
+   * @throws StorageException if preconditions fail
```

Comment:
I think `upon failure` is better than `if preconditions fail`, preconditions is subset of failure and we are bound to throw StorageException for other cases, either now or sometime in the future.

Reply:
'upon failure' is a formal phrase which means next to nothing to me. Idea of this message is tell user that method  either returns new blob or throws StorageException if preconditions fail.

Reply:
Ack. What are probabilities of runtime exception being wrapped inside `StorageException`? If its zero then we should be good.

Reply:
okay, you convinced me. changed to 'upon failure'

Reply:
I do like your idea of having specific phrases as much as possible but I have seen many times runtime exceptions or in future same exception being used to portray other scenarios in other methods(eg. currently this would fail for precondition checks in Storage#update, but it may represent some other exception in Storage#delete etc.).

So our aim should be not to be too specific about the generalized exception.

I hope that helps to convey my points.

Reply:
Thanks for your valuable comments, I really appreciate this.